rest-service/manager_rest/resource_manager.py

```diff
@@ -63,6 +63,65 @@
 from . import manager_exceptions
 
 
+class _SubDepSummary(object):
+    """A container for sub-counts and sub-statuses of a deployment.
+
+    This is only to be used by the recalc_ancestors machinery. This
+    class defines how counts and statuses should be added together.
+    """
+    __slots__ = (
+        'sub_services_count',
+        'sub_environments_count',
+        'sub_services_status',
+        'sub_environments_status'
+    )
+
+    def __init__(
+        self,
+        *,
+        sub_services_count=0,
+        sub_environments_count=0,
+        sub_services_status=None,
+        sub_environments_status=None,
+    ):
+        self.sub_services_count = sub_services_count
+        self.sub_environments_count = sub_environments_count
+        self.sub_services_status = sub_services_status
+        self.sub_environments_status = sub_environments_status
+
+    def __add__(self, other):
+        return self.__class__(
+            sub_services_count=self.sub_services_count +
+            other.sub_services_count,
+            sub_environments_count=self.sub_environments_count +
+            other.sub_environments_count,
+            sub_services_status=models.Deployment.compare_statuses(
+                self.sub_services_status, other.sub_services_status
+            ),
+            sub_environments_status=models.Deployment.compare_statuses(
+                self.sub_environments_status, other.sub_environments_status
+            )
+        )
+
+    def __iadd__(self, other):
```

Comment:
we could let python fall back to just `__add__`, but this is an optimization, it'll allow `a += b` to not create an additional object